README.md
[![Build Status](https://travis-ci.org/ynohat/akamai-g2o.svg?branch=master)](https://travis-ci.org/ynohat/akamai-g2o)

# Akamai G2O for Node JS

G2O (Ghost to Origin, aka Signature Header Authentication) is an Akamai protocol that provides request authentication
between Akamai caching proxies and the origin.

This library provides a simple middleware function that provides compliant G2O validation for applications. The following
NodeJS server frameworks are supported:

- vanilla NodeJS HTTP server
- Express 4
- Restify 4

## How the protocol works

From the documentation:

> This feature configures edge servers to include two special headers in requests to the origin server. One of these
> headers contains basic information about the specific request. The second header contains similar information encrypted
> with a shared secret. This allows the origin server to perform several levels of authentication to ensure that the request
> is coming directly from an edge server without tampering.

## What this library does not support

The reference documentation specifies that the `X-Akamai-G2O-Auth-Data` header should be stored to prevent replay attacks.
This is not supported (yet), because the library does not make assumptions about its environment and the storage engines
that may be available. It does however provide a way of extending the validation mechanism. See the `extraCheck` option
for more details.

## How to use this library

First, install it:

```
npm install --save akamai-g2o
```

It is assumed that you are familiar with G2O concepts, which aren't covered in depth here.

This simple example sets up an application to use G2O with default parameters.

```javascript
const app = require("express")();
const g2o = require("akamai-g2o");

app.use(g2o({
   key: {
    "nonce1": "s3cr3tk3y",
    "nonce2": "s3cr3tk3y2"
   }
}));

// ... setup routes and listen
```
Requests that fail authentication will result in a 403 response.

Requests that pass authentication will be extended with a `g2o` attribute with the following structure:

```javascript
{
    data: {}, // see the Data object description section for details
    signature: "base64 encoded signature generated by the middleware",
    authenticated: true|false,
    message: "present only when authenticated is false"
}
```

See the options section for more advanced configuration.

## Options

### key

Mandatory; `Object` or `Function`.

When providing an `Object`, it should map nonces to keys (see the official G2O documentation for more information about nonces).

It is also possible to provide a `Function` for more flexibility. It should have the following signature:

```javascript
function (req, data, callback)
```

`req` is the request object, typically an instance of `http.IncomingMessage`.

`data` is an `Object` representing the data in the `dataString`. It is described in more detail in the [Data object description](#data-object-description).

`callback` takes an `Error` or `null` as its first parameter and the key as its second.

### strict

*** NO LONGER SUPPORTED ***

If you want to implement not strict logic see: [onUnauthenticated](#onUnauthenticated).

### dataHeader

Optional; `String` (default: "X-Akamai-G2O-Auth-Data")

By default, Ghost sends the request data in the `X-Akamai-G2O-Auth-Data` header. Setting this option instructs the middleware
to retrieve a different header.

### signHeader

Optional; `String` (default: "X-Akamai-G2O-Auth-Sign")

By default, Ghost sends the request signature in the `X-Akamai-G2O-Auth-Sign` header. Setting this option instructs the middleware
to retrieve a different header.

### signString

Optional; `Function` (default: `req => req.url`)

By default, Ghost signs the request URL as represented in the status line of the request (the URL path).

Ghost can be configured to use any part or combination of parts of the request, in which case you
should provide a `Function` with the following signature:

```javascript
function (req) => String
```

### checkTime

Optional; `Boolean` (default: `true`)

If `true`, the request will fail if the request time is more than 30s distant from the current server time.

### timeWindow

Optional; `Number` (default: 30)

Number of seconds before or after which the difference between the server and request times will trigger an
authentication error when `checkTime` is `true`.

### extraCheck

Optional; `Function` (default: null)

If provided, the function will be called after all checks have been completed. The signature should be:

```javascript
function (req, callback)
```

An example implementation that prevents replay attacks might look like this:

```javascript
var g2o = require("akamai-g2o");
var app = require("express")();

var previousAuthDataValues = {};

app.use(g2o({
   key: {
    "nonce1": "s3cr3tk3y",
    "nonce2": "s3cr3tk3y2"
   },
   extraCheck: function (req, callback) {
       if (req.g2o.data.raw in previousAuthDataValues) {
           callback(new Error("replayed request"));
       } else {
           callback();
       }
   }
}));
```

### onUnauthenticated

Optional; `Function` (default: 

```javascript
function (req, res, next) {
    var statusCode = 403;
    if (typeof res.status === "function") {
        // Express has a status() helper function
        res.status(statusCode);
    } else {
        res.statusCode = statusCode;
    }
    res.end();
}
```

)

If provided, the function will be called after all checks have been completed and found to be unauthenticated.
An unauthenticated g2o data will have a message property, that is the failure reason. The signature should be:

```javascript
function (req, res, next)
```

An example implementation that:

- logs out the failed request
- uses a different status code
- only fails request if strict

```javascript
var g2o = require("akamai-g2o");
var app = require("express")();
var strict = true; // strict controlled outside, but maybe some config.

app.use(g2o({
   key: {
    "nonce1": "s3cr3tk3y",
    "nonce2": "s3cr3tk3y2",
   },
   onUnauthenticated: function (req, res, next) {
      var g2oResponse = Object.assign(
         {},
         {
            strict: strict,
            clientIp: req.ip, // for if there are no g2o header fallback to server known ip.
            forwardAddresses: req.forwardAddresses,
            uri: req.originalUrl, // note this is for express 4.0, else it is req.url
         },
         req.g2o
      );

      console.log('g2o unauthenticated', g2oResponse); // logging

      if (strict) { // only fail request if strict
         res.status(407).end(); // different status code
      } else {
         next();
      }
   }
}));
```

## Data object description

Ghost is required to send a header containing authentication information which is used to generate the signature. This library parses
the contents of the header and makes it available as an `Object` with the following structure:

```javascript
{
    // the value of the raw header
    raw: "1, 1.2.3.4, 2.3.4.5, 123456789, 42314563, nonce1",
    // specifies the hashing function to use
    version: int(1..5),
    // edge server IP address
    edgeIp: String,
    // client IP address
    clientIp: String,
    // request time, as a Date object
    time: Date,
    // request unique identifier
    uniqueId: String,
    // nonce referencing the key to use
    nonce: String
}
```

## Contributing

Don't break the unit tests, be as clean as you can, be nice.

```javascript
npm run -s test
```


CHANGELOG.md
Changelog
=========

## 2.0.0

### Remove Strict

Removed the strict option, and replaced with `onUnauthenticated` hook.

This update allows the calling application to implement additional functionality
when a request is found to be unauthenticated.  The examples shows:

- logging
- different status code
- reimplement strict

### Fail with 403 (\#2)

Previous versions incorrectly failed with 401.

## 1.0.0

Initial release.


test/sign.js
"use strict";

var yaml = require("js-yaml");
var fs = require("fs");
var sign = require("../lib/sign");

describe("signature algorithms", function () {
    var tests = yaml.safeLoad(fs.readFileSync(require.resolve("./sign/samples.yaml"), 'utf8'));

    Object.keys(tests).forEach(function (name) {
        var test = tests[name];

        it(name, function () {
            sign(
                test.version,
                test.key,
                test.authData,
                test.signString
            ).should.equal(test.signature);
        });
    });
});


test/index.js
"use strict";

var yaml = require("js-yaml");
var fs = require("fs");

describe("akamai-g2o", function () {
    require("./sign");
    require("./middleware");
});


test/middleware.js
"use strict";

describe("middleware", function () {
    require("./middleware/raw");
    require("./middleware/express");
    require("./middleware/restify");
});


lib/sign.js
"use strict";

var crypto = require("crypto");

/**
 * Generates a request signature.
 *
 * @throws {Error} If the version is not implemented
 * @param  {Number} version the signing algorithm version
 * @param  {String} key the shared signing key
 * @param  {String} data the contents of the data header (typically X-Akamai-G2O-Data)
 * @param  {String} signString typically the request URL
 * @return {String} the signature
 */
module.exports = function (version, key, data, signString) {
    // versions start at 1, decrement to fit in the array
    version--;
    if (version in versions) {
        return versions[version](key, data, signString);
    }
    throw new Error("unsupported algorithm");
};

var versions = [
    // version 1: MD5(key,data,sign-string)
    function (key, data, signString, encoding) {
        var h = crypto.createHash("md5");
        h.update(key + data + signString);
        return h.digest(encoding || "base64");
    },

    // version 2: MD5(key,MD5(key,data,sign-string))
    function (key, data, signString, encoding) {
        var h1 = crypto.createHash("md5");
        h1.update(key);
        h1.update(data);
        h1.update(signString);
        var h2 = crypto.createHash("md5");
        h2.update(key);
        h2.update(h1.digest());
        return h2.digest(encoding || "base64");
    },

    // version 3: MD5-HMAC(key,data,sign-string)
    function (key, data, signString, encoding) {
        var h = crypto.createHmac("md5", key);
        h.update(data);
        h.update(signString);
        return h.digest(encoding || "base64");
    },

    // version 4: SHA1-HMAC(key,data,sign-string)
    function (key, data, signString, encoding) {
        var h = crypto.createHmac("sha1", key);
        h.update(data);
        h.update(signString);
        return h.digest(encoding || "base64");
    },

    // version 5: SHA256-HMAC(key,data,sign-string)
    function (key, data, signString, encoding) {
        var h = crypto.createHmac("sha256", key);
        h.update(data);
        h.update(signString);
        return h.digest(encoding || "base64");
    }
];


lib/middleware.js
"use strict";

var sign = require("./sign");

/**
 * Returns a middleware function using the provided options.
 *
 * @param  {Object} options
 * @return {Function}
 */
module.exports = function (options) {
    options = sanitizeOptions(options);
    return function (req, res, next) {
        req.g2o = {};
        req.g2o.authenticated = true;
        authenticateRequest(options, req, function (error) {
            if (error) {
                req.g2o.authenticated = false;
                req.g2o.message = error.message;
            }
            if (!req.g2o.authenticated) {
                options.onUnauthenticated(req, res, next);
            } else {
                next();
            }
        });
    };
};

function authenticateRequest(options, req, callback) {
    var dataHeader = options.dataHeader.toLowerCase();
    if (!(dataHeader in req.headers)) {
        return callback(new Error("missing auth data header"));
    }
    var dataString = req.headers[dataHeader];
    var data = parseDataString(dataString);
    req.g2o.data = data;
    checkTime(data, options, req, function (error) {
        if (error) return callback(error);
        checkSignature(data, options, req, function (error) {
            if (error) return callback(error);
            if (typeof options.extraCheck === "function") {
                options.extraCheck(req, callback);
            } else {
                callback();
            }
        });
    });
}

function checkTime(data, options, req, callback) {
    if (options.checkTime) {
        if (Math.abs(Date.now() - data.time.getTime()) > 1000 * options.timeWindow) {
            return callback(new Error("request time too far off"));
        }
    }
    callback();
}

function checkSignature(data, options, req, callback) {
    var signHeaderName = options.signHeader.toLowerCase();
    if (!(signHeaderName in req.headers)) {
        return callback(new Error("missing sign header"));
    }
    var signString = options.signString(req);
    options.key(req, data, function (error, key) {
        if (error) return callback(error);
        try {
            var signature = sign(data.version, key, data.raw, signString);
            var ghostSignature = req.headers[signHeaderName];
            if (signature !== ghostSignature) {
                return callback(new Error("invalid signature"));
            }
            callback();
        } catch (error) {
            callback(error);
        }
    });
}

function parseDataString(dataString) {
    // X-Akamai-G2O-Auth-Data: version, edge-ip, client-ip, time, unique-id, nonce
    var bits = dataString.split(", ");
    var data = {};
    data.raw = dataString;
    data.version = Number(bits.shift());
    data.edgeIp = bits.shift();
    data.clientIp = bits.shift();
    data.time = new Date(Number(bits.shift()) * 1000);
    data.uniqueId = bits.shift();
    data.nonce = bits.shift();
    return data;
}

function sanitizeOptions(options) {
    // create a copy and assign some sane defaults
    options = Object.assign({}, {
        key: null,
        checkTime: true,
        timeWindow: 30,
        dataHeader: "X-Akamai-G2O-Auth-Data",
        signHeader: "X-Akamai-G2O-Auth-Sign",
        signString: defaultSignStringFunction,
        extraCheck: null,
        onUnauthenticated: defaultOnUnauthenticatedFunction,
    }, options);

    // make sure the key option is a function even if the user
    // provided an object
    switch (typeof options.key) {
        case "object":
            options.key = function (req, data, callback) {
                if (data.nonce in this) {
                    return callback(null, this[data.nonce]);
                }
                callback(new Error("unknown nonce"));
            }.bind(options.key);
            break;
        case "function":
            break;
        default:
            throw new Error("key should be an object or a function");
    }

    return options;
}

function defaultSignStringFunction(req) {
    if (req.originalUrl) {
        var bits = require("url").parse(req.originalUrl);
        return bits.path;
    }
    return req.url;
}

function defaultKeyFunction(req, data, callback) {
    if (data.nonce in this) {
        return callback(null, this[data.nonce]);
    }
    callback(new Error("unknown nonce"));
}

function defaultOnUnauthenticatedFunction(req, res, next) {
    var statusCode = 403;
    if (typeof res.status === "function") {
        // Express has a status() helper function
        res.status(statusCode);
    } else {
        res.statusCode = statusCode;
    }
    res.end();
}
